robot/lib/Src/drivers/ICM20948.cpp

```diff
@@ -0,0 +1,180 @@
+    constexpr uint8_t USER_CTRL     = 0x03;
+    constexpr uint8_t LP_CONFIG     = 0x05; 
+    constexpr uint8_t PWR_MGMT_1    = 0x06;
+    constexpr uint8_t PWR_MGMT_2    = 0x07;
+
+    constexpr uint8_t GYRO_Z_OUT_H  = 0x37;
+    constexpr uint8_t GYRO_Z_OUT_L  = 0x38;
+
+    constexpr uint8_t FIFO_MODE     = 0x69;
+   
+    // Bank 2 
+    constexpr uint8_t GYRO_CONFIG_1 = 0x01;
+
+    enum GYRO_FS_SEL {
+        DPS_2000 = 3,
+        DPS_1000 = 2,
+        DPS_500 = 1,
+        DPS_250 = 0,
+    };
+
+    enum PWR_MGMT {
+        DEVICE_RESET = 1,
+        SLEEP = 1,  // 1: enter sleep mode
+        CLKSEL = 1, // 1-5: Auto selects best available clock source
+    };
+
+    enum LP_CONFIG_REG {
+        I2C_MST_CYCLE = 1,
+    };
+
+    enum USER_CTRL_REG {
+        I2C_IF_DIS = 1, // disable I2C, enable SPI
+    }; 
+    
+};
+
+ICM20948::ICM20948(std::shared_ptr<SPI> imuSPI, PinName cs_pin) 
+    : imuSPI(std::move(imuSPI)),
+      nCs(cs_pin) {
+    //chip_select_pin.write(false);
+    nCs = 1;
+    HAL_Delay(1500);
+}
+
+void ICM20948::initialize() {
+    // SPI pulse to clear pins. 
+    imuSPI->transmit(0);    
+    
+    uint8_t whoami = read_register(0, Registers::WHO_AM_I);
+    
+    while (whoami != WHOAMI_VAL) {
+        printf("Failed to connect to IMU.");
+        whoami = read_register(Registers::WHO_AM_I);
+    }
+    
+    // Reset
+    write_register( 
+        Registers::PWR_MGMT_1,
+        Registers::PWR_MGMT::DEVICE_RESET << 7
+        | Registers::PWR_MGMT::SLEEP << 6
+        | Registers::PWR_MGMT::CLKSEL);
+
+    HAL_Delay(100);
+
+    // Check whoami again
+    whoami = read_register(0, Registers::WHO_AM_I);
+    
+    while (whoami != WHOAMI_VAL) {
+        printf("Failed to connect to IMU.");
+        whoami = read_register(Registers::WHO_AM_I);
+    }
+    
+    // Auto select best clock source
+    write_register(
+        Registers::PWR_MGMT_1, 
+        Registers::PWR_MGMT::CLKSEL);
+    
+    // Operate I2C in duty cycled mode and disable GYRO duty cycle mode
+    write_register(
+        Registers::LP_CONFIG, 
+        Registers::LP_CONFIG_REG::I2C_MST_CYCLE << 6);
+    
+    // Disable I2C
+    write_register(
+        Registers::USER_CTRL,
+        Registers::USER_CTRL_REG::I2C_IF_DIS << 4);
+
+    // Turn FIFO off
+    write_register(Registers::FIFO_MODE, 0x00);
+
+    // Enable gyro
+    write_register(Registers::PWR_MGMT_2, 0x00);
+    
+    // Configure gyro
+    write_register(
+        2, 
+        Registers::GYRO_CONFIG_1, 
+        Registers::GYRO_FS_SEL::DPS_1000 << 2);
+    
+    
+    printf("Initialization Complete!");
+}
+
+float ICM20948::gyro_z() {
+
+    uint16_t hi = read_register(0, Registers::GYRO_Z_OUT_H),
+             lo = read_register(0, Registers::GYRO_Z_OUT_L);
+
+
+    // Reinterpret the bits as a signed 16-bit integer
+    int16_t total = hi << 8 | lo;
+
+    // +-1000dps max, for 16-bit signed integers, comes out to 32.8lsb/deg.
+    const float lsbToDps = 1000.0f / 32768.f;
```

Comment:
The data sheet almost always describes the scale factor to go from LSB to deg/sec. See table 1 in section 3.1. I have seen some devices in which the output of the gyro doesn't actually fill all 16 bits resulting in super incorrect numbers. In this case it's ok, but it's good to keep it in mind for the future.  The conversion is done following the instructions found in 8.25.

![image](https://user-images.githubusercontent.com/5797304/75078528-74a0a200-54cb-11ea-91eb-cee3d3d00a72.png)
![image](https://user-images.githubusercontent.com/5797304/75078581-97cb5180-54cb-11ea-9a4d-caa930785618.png)

Also, these constants can be constexpr. It doesn't make too much of difference in this case, but it does the calculations at compile time instead of possibly during runtime.

Reply:
This matches the data sheet (and the data sheet actually rounds in an undesirable way), so no worries in this case - I think it's preferable as-is.

I mean, this will almost certainly be optimized out at compile-time regardless. But yeah, it never hurts to make it constexpr.

Reply:
Due to how the chip is made, I don't think we can actually say if the data sheet rounds the scale factor number or the full-range number. Either way, the +-3% tolerance on the scale factor greatly outweighs this rounding anyways so it's not a big deal at all.

I was mostly throwing it out there for info that this exists in the data sheet and can easily cause major problems. I think two years ago someone did this exact same calculation and it was a few orders of magnitude off resulting it horribly incorrect numbers